posts/2018-12-21-Procedural-Macros-in-Rust-2018.md

```diff
@@ -0,0 +1,363 @@
+involved with them because the 2018 edition has so dramatically improved the
+experience both defining and using them.
+
+Here I'd like to explore what procedural macros are, what they're capable of,
+notable new features, and some fun use cases of procedural macros. I might even
+convince you that this is Rust 2018's best feature as well!
+
+### What is a procedural macro?
+
+First defined over two years ago in [RFC 1566], procedural macros are, in
+layman's terms, a function that takes a piece of syntax at compile time and
+produces a new bit of syntax. Procedural macros in Rust 2018 come in one of
+three flavors:
+
+* **`#[derive]` mode macros** have actually been stable since [Rust 1.15]
+  and bring all the goodness and ease of use of `#[derive(Debug)]` to
+  user-defined traits as well, such as [Serde]'s `#[derive(Deserialize)]`.
+
+* **Function-like macros** are new to the 2018 edition and allow defining
+  macros like `env!("FOO")` or `format_args!("...")` in a crate.io-based
+  library. You can think of these as sort of "`macro_rules!` macros" on
+  steroids.
+
+* **Attribute macros**, my favorite, are also new in the 2018 edition
+  and allow you to provide lightweight annotations on Rust functions which
+  perform syntactical transformations over the code at compile time.
+
+Each of these flavors of macros can be defined in a crate with `proc-macro =
+true` [specified in its manifest][manifest]. When used, a procedural macro is
+loaded by the Rust compiler and executed as the invocation is expanded. This
+means that Cargo's in control of versioning for procedural macros and you can
+use them with all same ease of use you'd expect from other Cargo dependencies!
+
+### Defining a procedural macro
+
+Each of the three types of procedural macros are [defined in a slightly different
+fashion][proc-ref], and here we'll single out attribute macros. First we'll flag
+`Cargo.toml`:
+
+```toml
+[lib]
+proc-macro = true
+```
+
+and then in `src/lib.rs` we can write our macro:
+
+```rust
+extern crate proc_macro;
+use proc_macro::TokenStream;
+
+#[proc_macro_attribute]
+pub fn hello(attr: TokenStream, item: TokenStream) -> TokenStream {
+    // ...
+}
+```
+
+We can then write some unit tests in `tests/smoke.rs`:
+
+```rust
+#[my_crate::hello]
+fn wrapped_function() {}
+
+#[test]
+fn works() {
+    wrapped_function();
+}
+```
+
+... and that's it! When we execute `cargo test` Cargo will compile our
+procedural macro. Afterwards it will compile our unit test which loads the macro
+at compile time, executing the `hello` function and compiling the resulting
+syntax.
+
+Right off the bat we can see a few important properties of procedural macros:
+
+* The input/output is this fancy `TokenStream` type we'll talk about more in a
+  bit
+* We're *executing aribtrary code* at compile time, which means we can do just
+  about anything!
+* Procedural macros are incorporated with the module system, meaning no more
+  janky `#[macro_use]`
```

Comment:
I would phrase this positively rather than negatively, e.g., "meaning they can be used just like functions and types"